gytheio-content-handlers/gytheio-content-handler-s3/src/test/java/org/gytheio/content/handler/s3/S3ContentReferenceHandlerImplTest.java

```diff
@@ -202,12 +216,133 @@ private void testFileOpsImpl()
         }
         catch (IOException ioe)
         {
-        	System.err.println(ioe.getMessage());
+        	logger.error(ioe.getMessage());
+        }
+
+        if (cleanup)
+        {
+            // delete the S3 object
+            handler.delete(reference);
+
+            assertFalse(handler.isContentReferenceExists(reference));
         }
+    }
+    
+    /**
+     * Tests: putFile (+ delete for cleanup)
+     */
+    @Test
+    public void testPutFile() throws IOException
+    {
+        String envVar;
+
+        envVar = getVar("pf_start_size");
+        long pf_start_size = (envVar != null ? new Long(envVar) : 1024 * 1024 * 1L); // default 1 MiB
+
+        envVar = getVar("pf_multiplier");
+        int pf_multiplier = (envVar != null ? new Integer(envVar) : 2);
 
-        // delete the S3 object
-        handler.delete(reference);
+        envVar = getVar("pf_count");
+        int pf_count = (envVar != null ? new Integer(envVar) : 3);
+
+        envVar = getVar("pf_repeat");
+        int pf_repeat = (envVar != null ? new Integer(envVar) : 1);
+
+        if (logger.isInfoEnabled())
+        {
+            logger.info("testPutFile: pf_start_size="+pf_start_size+", pf_multiplier="+pf_multiplier
+                    +", pf_count="+pf_count+", pf_repeat="+pf_repeat);
+        }
+
+        long testContentSizeInBytes = pf_start_size;
+
+        for (int i = 0; i < pf_count; i++)
+        {
+            File testFile = createTestFile(testContentSizeInBytes);
+
+            for (int j = 0; j < pf_repeat; j++)
+            {
+                // via S3 Transfer Manger Upload
+                testPutFileImpl(testFile, true);
+
+                // via S3 PutObject
+                testPutFileImpl(testFile, false);
+            }
+
+            if (cleanup)
+            {
+                testFile.delete();
+            }
+
+            // increase test file size for next loop
+            testContentSizeInBytes = testContentSizeInBytes * pf_multiplier;
+        }
+    }
+
+    private void testPutFileImpl(File testFile, boolean usePutFile) throws FileNotFoundException
+    {
+        String uuid = UUID.randomUUID().toString();
+
+        String fileName = "test-" + uuid + ".bin";
+
+        ContentReference reference = handler.createContentReference(fileName, "application/octet-stream");
 
         assertFalse(handler.isContentReferenceExists(reference));
+
+        long contentLen = testFile.length();
+
+        reference.setSize(contentLen); // prevents AmazonS3Client warn
```

Comment:
Maybe rename `createTestFile` to `createTestFileContentReference`, have it return a `ContentReference` and pack all of this in there?

Reply:
FYI: The "reference" here was the target ContentReference. 

Not sure if the source test file really needs to be a Gytheio ContentReference (but could be) ? It might be more useful if we ever need to add additional S3 tests &/or create generic (abstract) test code across different content handlers ?

For now, I've renamed slightly to make it clearer. Happy to re-work further, if needed.